examples/cpp/pyperf/PyPerfNativeStackTrace.cc

```diff
@@ -191,5 +219,95 @@ bool NativeStackTrace::error_occured() const {
   return error_occurred;
 }
 
+bool NativeStackTrace::is_cached(const MAP &map, const uint32_t &key) {
+  try {
+      map.at(key);
+      return true;
+  }
+  catch (const std::out_of_range&) {
+      logInfo(2, "No entry for %d in the cache\n", key);
+  }
+  return false;
+}
+
+Object NativeStackTrace::cache_get(const MAP &map, const uint32_t &key) {
+  const Object & value = map.at(key);
+  return value;
+}
+
+// cache_put adds a new entry to the map if the capacity allows
+void NativeStackTrace::cache_put(MAP &map, const uint32_t &key, const unw_cursor_t cursor, const unw_addr_space_t as, void *upt) {
+  
+  // Check available capacity
+  if (cache_size() > NativeStackTrace::CacheMaxSizeMB*1024*1024 - cache_single_entry_size()) { 
+    logInfo(2, "The cache usage is %.2f MB, close to reaching the max memory usage (%d MB)\n", cache_size_KB()/1024, NativeStackTrace::CacheMaxSizeMB);
+    logInfo(2, "Skipping adding an entry for %d to the cache\n", key);
+     
+    return;
+  }
+
+  Object obj = {cursor, as, upt, time(NULL)};
+  map[key] = obj;
+  logInfo(2, "New entry for %d was added to the cache\n", key);
+}
+
+// cache_delete_key removes the element from the cache and destroys unwind address space and UPT
+// to ensure that all memory and other resources are freed up
+bool NativeStackTrace::cache_delete_key(MAP &map, const uint32_t &key) {
+  try {
+    Object v = cache_get(map, key);    
+    map.erase(key);
+    cleanup(v.upt, v.as);
+    logInfo(2, "The entry for %d was deleted from the cache\n", key);
+    return true;
+  }
+  catch (const std::out_of_range&) {
```

Comment:
`try`-`catch` scope can be limited to the `cache_get` call (and not to `map.erase` and `cleanup` calls)

Reply:
Moved `map.erase` and `cleanup` outside the `try`-`catch`